driver-core/src/test/unit/com/mongodb/connection/CommandMessageSpecification.groovy

```diff
@@ -152,8 +152,92 @@ class CommandMessageSpecification extends Specification {
+        def sessionContext = Stub(SessionContext)
+
+        when:
+        message.encode(output, sessionContext)
+        def byteBuf = new ByteBufNIO(ByteBuffer.wrap(output.toByteArray()))
+        def messageHeader = new MessageHeader(byteBuf, maxMessageSize)
+
+        then:
+        messageHeader.opCode == OpCode.OP_MSG.value
+        messageHeader.requestId < RequestMessage.currentGlobalId
+        messageHeader.responseTo == 0
+        messageHeader.messageLength == 1024
+        byteBuf.getInt() == 0
+        payload.getPosition() == 1
+        payload.hasAnotherSplit()
+
+        when:
+        payload = payload.getNextSplit()
+        message = new CommandMessage(namespace, command, fieldNameValidator, ReadPreference.primary(), messageSettings,
+                false, payload, fieldNameValidator)
+        output.truncateToPosition(0)
+        message.encode(output, sessionContext)
+        byteBuf = new ByteBufNIO(ByteBuffer.wrap(output.toByteArray()))
+        messageHeader = new MessageHeader(byteBuf, maxMessageSize)
+
+        then:
+        messageHeader.opCode == OpCode.OP_MSG.value
+        messageHeader.requestId < RequestMessage.currentGlobalId
+        messageHeader.responseTo == 0
+        messageHeader.messageLength == 1024
+        byteBuf.getInt() == 0
+        payload.getPosition() == 2
+        payload.hasAnotherSplit()
+
+        when:
+        payload = payload.getNextSplit()
+        message = new CommandMessage(namespace, command, fieldNameValidator, ReadPreference.primary(), messageSettings,
+                false, payload, fieldNameValidator)
+        output.truncateToPosition(0)
+        message.encode(output, sessionContext)
+        byteBuf = new ByteBufNIO(ByteBuffer.wrap(output.toByteArray()))
+        messageHeader = new MessageHeader(byteBuf, maxMessageSize)
+
+        then:
+        messageHeader.opCode == OpCode.OP_MSG.value
+        messageHeader.requestId < RequestMessage.currentGlobalId
+        messageHeader.responseTo == 0
+        messageHeader.messageLength == 552
+        byteBuf.getInt() == 0
+        payload.getPosition() == 1
+        payload.hasAnotherSplit()
+
+        when:
+        payload = payload.getNextSplit()
+        message = new CommandMessage(namespace, command, fieldNameValidator, ReadPreference.primary(), messageSettings,
+                false, payload, fieldNameValidator)
+        output.truncateToPosition(0)
+        message.encode(output, sessionContext)
+        byteBuf = new ByteBufNIO(ByteBuffer.wrap(output.toByteArray()))
+        messageHeader = new MessageHeader(byteBuf, maxMessageSize)
+
+        then:
+        messageHeader.opCode == OpCode.OP_MSG.value
+        messageHeader.requestId < RequestMessage.currentGlobalId
+        messageHeader.responseTo == 0
+        messageHeader.messageLength == 562
+        byteBuf.getInt() == 1 << 1
+        payload.getPosition() == 1
+        !payload.hasAnotherSplit()
+    }
+
+    def 'should respect the max batch count'() {
```

Comment:
Split the max batch count assertions into a separate test as the needs of the two tests diverged sufficiently to make it not worth it to keep them together.